### PR TITLE
doc: boards: secure and non-secure domain defs

### PR DESCRIPTION
**Pull request by greg-fer**
* Added definitions of secure and non-secure domain. Updated board support and glossary.
   NCSDK-17496.
* Replaced "secure domain" and "non-secure domain" with SPE and NSPE.
   NCSDK-15298.

Signed-off-by: Grzegorz Ferenc <Grzegorz.Ferenc@nordicsemi.no>

----

**NOTE:** sdk-nrfxlib edits will be done as part of separate task.

### CODE REVIEW

**Comment by greg-fer** [comment]
@mia-ko , @joerchan , this is a task that adds the definitions @trond-snekvik pointed out were missing in the docs.
I am aware of the discussion around NCSDK-15298, but this task is just broadly related to it.

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:81]: The build does not seem to like the asterisk at this spot. Would it work adding a backslash just before the asterisk, like this \*?
Or enclosing it in double back-quotes (inline literals), like this ``*``.

Do we need this though? It feels a bit like a footnote :)

**Review by mia-ko — CHANGES_REQUESTED** [review]

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:27]: Much cleaner solution :)

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:145]:
```suggestion
Secure domain and non-secure domain are two concepts related to the way of organizing the firmware in the application core of a multi-core SoC, when the firmware is programmed to the ``cpuapp`` or ``cpuapp_ns`` build targets.
```

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:154]: Consider adding a blank line before/splitting to new paragraph.

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:157]: If we would like to have this information here about the SPM deprecation, then maybe remove the brackets.

File: doc/nrf/glossary.rst
Comment (mia-ko) [inline doc/nrf/glossary.rst:577]:
```suggestion
      * When there is no non-secure domain: All firmware components.
```

File: doc/nrf/ug_nrf5340.rst
Comment (mia-ko) [inline doc/nrf/ug_nrf5340.rst:61]:
```suggestion
* ``nrf5340dk_nrf5340_cpuapp`` for the secure domain only.
```

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Review by trond-snekvik — APPROVED** [review]
LGTM, but someone from the TF-M team (@joerchan?) should have a look.

File: doc/nrf/ug_nrf5340.rst
Comment (mia-ko) [inline doc/nrf/ug_nrf5340.rst:62]: Not sure if I feel that "adding" sits well with me. It's just a feeling, without me being completely able to say why, so I will leave this completely up to you to decide if you want to change or not. 
The alternative would maybe be to write "for both the secure and non-secure domain

File: doc/nrf/ug_nrf9160.rst
Comment (mia-ko) [inline doc/nrf/ug_nrf9160.rst:179]: I think we should maybe either mention firmware in this introductory sentence, or keep firmware in the bullet points :)

**Review by mia-ko — CHANGES_REQUESTED** [review]

File: doc/nrf/app_boards.rst
Comment (mia-ko) [inline doc/nrf/app_boards.rst:182]: "On selecting the build target for NSPE" is contradicting the sentence just before a little bit, in my opinion. I think we might just write "On selecting the ``_ns`` build target" and leave the rest of the sentence as is. Or:
```suggestion
On selecting the ``_ns`` build target, the application is built as a non-secure image to reside in NSPE, and an additional firmware component is built as the secure image to reside in SPE.
```

**Review by mia-ko — CHANGES_REQUESTED** [review]
Good stuff, Greg, this really starts to look good IMO :)
Just one minor suggestion for change from my side, see if you agree.

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst]: This section was re-written.

Comment (mia-ko) [reply doc/nrf/app_boards.rst]: Perfect!

**Comment by greg-fer** [comment]
@joerchan , @hakonfam , can you have a look sometime this week?

**Review by lemrey — APPROVED** [review]

**Review by mia-ko — CHANGES_REQUESTED** [review]
:)
To me, this is the best explanation and structure so far.
Just have a few comments now.
Also, we need to update the build_and_run_ns.txt file as well.

File: doc/nrf/glossary.rst
Comment (hakonfam) [inline doc/nrf/glossary.rst:415]:
```suggestion
      In the Security by Separation scenario, NSPE refers to the firmware components located in the non-secure partition, such as main user application, while all the security components reside in the :term:`Secure Processing Environment (SPE)`.
```

Does this correspond with the term used elsewhere in this PR?

File: doc/nrf/libraries/others/secure_services.rst
Comment (hakonfam) [inline doc/nrf/libraries/others/secure_services.rst:10]: Any reason we don't use abbreviations here for SPE and NSPE?

File: doc/nrf/templates/sample_README.rst
Comment (hakonfam) [inline doc/nrf/templates/sample_README.rst:178]:
```suggestion
   * Include the standard text for building - either ``.. include:: /includes/build_and_run.txt``, or ``.. include:: /includes/build_and_run_ns.txt`` for the build targets that include Non-Secure Processing Environment (NSPE).
```
?

File: doc/nrf/ug_nrf91_features.rst
Comment (hakonfam) [inline doc/nrf/ug_nrf91_features.rst:82]: `user application` 
`main application`
`main user application` 

I think all these three terms are used for the same concept, we should consider aligning.

File: doc/nrf/app_boards.rst
Comment (frkv) [reply doc/nrf/app_boards.rst:23]: I concur. I would rather have this described more accurately describing that the application is in Non-Secure Processing environment (NSPE), and that this relies Secure Processing Environment for security features

File: doc/nrf/app_boards.rst
Comment (frkv) [reply doc/nrf/app_boards.rst:147]: Concur: The Secure Processing Environment (and the Arm PSA spec) is already using the term partition when it comes to separation e.g. of ARoT services and PRoT services as well as select services inside the Platform Root of Trust (e.g. PSA crypto service, PSA attestation service and PSA ITS) as being "partitions"

Environment is much better!

File: doc/nrf/app_boards.rst
Comment (frkv) [reply doc/nrf/app_boards.rst:175]: I like the addition. It reminds the user of this detail in clear terms

File: doc/nrf/app_boards.rst
Comment (frkv) [reply doc/nrf/app_boards.rst:177]: I don't think that we should talk about security-critical code in the NSPE, but the typically is better as it shows that you can do multiple things in that scope.

I propose we make this change

File: doc/nrf/app_boards.rst
Comment (mia-ko) [reply doc/nrf/app_boards.rst:147]: Oooh, I like the second one here, it makes sense. But don't drop the comma ;)

File: doc/nrf/app_boards.rst
Comment (frkv) [inline doc/nrf/app_boards.rst:22]: Secure Processing Environment as a concept is related to the usage of ArmV8-M Security Extension. I don't think it is completely accurate to call this Secure Processing Environment (SPE)

```suggestion
* ``cpuapp`` -- When you choose this target, you build the application core firmware in a single execution environment, that does not use :ref:`Trusted Firmware-M (TF-M) <ug_tfm>`.
```

File: doc/nrf/ug_nrf7002.rst
Comment (greg-fer) [reply doc/nrf/ug_nrf7002.rst:184]: Right.

**Comment by greg-fer** [comment]
@mia-ko , addressed your comments. Working now on comments from the rest of reviewers.

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:23]: Rephrased to:
```
* ``cpuapp`` -- When you choose this target, you build the application core firmware that is entirely in the Secure Processing Environment (SPE) and does not use Cortex-M Security Extensions (:ref:`Trusted Firmware-M (TF-M) <ug_tfm>`).
* ``cpuapp_ns`` -- When you choose this target, you build the application with Cortex-M Security Extensions (CMSE).
  The application core firmware is placed in Non-Secure Processing Environment (NSPE) and uses Secure Processing Environment (SPE) for security features.
  The build system automatically includes :ref:`Trusted Firmware-M (TF-M) <ug_tfm>` in SPE and merges it with NSPE.
```

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:147]: Rephrased to "two security environments (flash partitions)"

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:173]: Rephrased to "CMSE is enabled and there is the following separation between the two environments:".

File: doc/nrf/glossary.rst
Comment (greg-fer) [reply doc/nrf/glossary.rst:415]: Rephrased both definitions in the glossary and added CMSE term.

File: doc/nrf/libraries/others/secure_services.rst
Comment (greg-fer) [reply doc/nrf/libraries/others/secure_services.rst:10]: The reason was more trivial: with the SPM link being in parentheses, I had no easy way to add the abbreviations without adding additional information.

Comment (greg-fer) [reply doc/nrf/libraries/others/secure_services.rst:10]: Rephrased to: "Secure Services are functions implemented in the firmware in Secure Processing Environment (SPE), but made available to be called from firmware in Non-Secure Processing Environment (NSPE)."

The link to SPM is provided in the next paragraph and in the API section, so I don't think we really need it in the first sentence.

File: doc/nrf/templates/sample_README.rst
Comment (greg-fer) [reply doc/nrf/templates/sample_README.rst:178]: Rephrased to: "for the build targets that use :ref:`Cortex-M Security Extensions <app_boards_spe_nspe>`."

File: doc/nrf/ug_nrf91_features.rst
Comment (greg-fer) [reply doc/nrf/ug_nrf91_features.rst:82]: Aligned to "user application".

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:22]: Rephrased to "* ``cpuapp`` -- When you choose this target, you build the application core firmware as a single execution environment that does not use Cortex-M Security Extensions (:ref:`Trusted Firmware-M (TF-M) <ug_tfm>`)."

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:20]: Rephrased into more bullet points. Please recheck after my push.

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:23]: This is now rephrased with earlier comments from Håkon and Frank Audun.

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:146]: Indeed, with your previous comment, I see where the confusion was. This is now rephrased to "The ``_ns`` suffix in the board target name refers to the usage of Cortex-M Security Extensions (CMSE)."

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:147]: Rephrased to "With CMSE, the firmware is stored in one of two security environments (flash partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE)."

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:145]: Addressed:
"The build target column in the tables above separates entries according to the CPU to target (for multi-core SoCs) and whether Cortex-M Security Extensions (CMSE) are used or not (addition of ``_ns`` if they are used)."

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:160]: This is now rephrased following the suggestion of adding CMSE as the differentiator. I hope this will now read better.

File: doc/nrf/app_boards.rst
Comment (greg-fer) [reply doc/nrf/app_boards.rst:173]: Addressed.

File: samples/nrf9160/http_update/full_modem_update/README.rst
Comment (greg-fer) [inline samples/nrf9160/http_update/full_modem_update/README.rst:84]: @joerchan , should this SPM mention be replaced with TFM? I think it might've been overlooked during the deprecation of SPM a release ago.

**Comment by greg-fer** [comment]
@hakonfam , @frkv , @joerchan , edits applied, please recheck.

**Review by hakonfam — APPROVED** [review]

File: doc/nrf/app_build_system.rst
Comment (frkv) [inline doc/nrf/app_build_system.rst:117]: What about a more normalized name "image" instead of "domain". The first line already uses "all images" in the documentation

```suggestion
|                                 | multi-image builds. In multi-core builds, several :file:`merged_<domain_name>.hex` fields       |                                                                                                                 |
```

File: samples/nrf9160/http_update/full_modem_update/README.rst
Comment (mia-ko) [reply samples/nrf9160/http_update/full_modem_update/README.rst:84]: This sentence feels redundant, because of the .. include:: /includes/build_and_run_ns.txt a few lines above.

File: doc/nrf/app_build_system.rst
Comment (greg-fer) [reply doc/nrf/app_build_system.rst]: Applied.

**Review by frkv — APPROVED** [review]
LGTM

**Review by mia-ko — CHANGES_REQUESTED** [review]
Looking a bit at the image (spe_nspe.svg)...
It's really a good visualization.
The only thing I am not sure about is if user application should be stated explicitly/make it visible in the NSPE/SPE box when using CMSE.
Or is what we have self-explanatory enough?
I'll leave it up to you to decide :)

**Review by mia-ko — APPROVED** [review]

File: doc/nrf/ug_nrf7002.rst
Comment (richabp) [inline doc/nrf/ug_nrf7002.rst:175]:
```suggestion
Applications on nRF5340 can use Cortex-M Security Extensions (CMSE) and separate firmware for the application core between Secure Processing Environment (SPE) and Non-Secure Processing Environment (NSPE).
```

**Review by richabp — CHANGES_REQUESTED** [review]

**Comment by greg-fer** [comment]
@richabp , applied your edits, please check.

**Review by richabp — APPROVED** [review]

**Comment by greg-fer** [comment]
@joerchan , going to merge this one, but if you have any feedback, let me know, and I will address it in the manifest PR for the PR that fixes sdk-nrfxlib occurrences of secure/non-secure domain.

**Pull request merged**
